### ramo else para for - 3.2.12

In [5]:

for i in range(1,5):
    if i==3: break # pode ser 1 linha desde que 1 instrução
    print("else:", i)
else: print("else:",i) # Só não passa se ocorrer break no laço    



else: 1
else: 2



### 3.2.14   LAB   Essenciais do loop while



In [10]:
blocks = int(input("Insira o número de blocos:"))  
 # Escreva seu código aqui.
altura = 0
total_blocos = 0
while total_blocos < blocks:
    altura += 1
    total_blocos += altura 



#

print("A altura da pirâmide:", altura)


A altura da pirâmide: 2


### 
Fazer um programa que solicite um numero e retorno a raiz
aproximada do numero.
Você deve usar laço for, if e break



In [28]:
num = int(input("Digite o valor"))
raiz = 0 
for i in range (num):
# digite seu codigo
   raiz = i
   if (raiz + 1) *(raiz +1 ) > num:
         break;
print("raiz aproximada é: ", raiz)                   

raiz aproximada é:  3


In [36]:
# teste CPF
cpf = input("Digite o cpf (Somente numero, 11 digitos): !")
cpf_sem_dg = ""
soma=0 
for i in range(9):
        soma += int(cpf[i]) * (10 -1)
        cpf_sem_dg += cpf[1]
# calculo primeiro difito
dg1 =soma * 10 % 11
if dg1 > 9: dg1 = 0
soma = 0
for i in range(9):
        soma += int(cpf[i]) * (11 -1)
# calculo digito 2 
dg2 = soma * 10 % 11
if dg2 > 9: dg2 = 0
# Montando cpf calculado
cpf_sem_dg += str(dg1) + str(dg2)
if cpf != cpf_sem_dg:
    print("Cpf invalido", cpf, cpf_sem_dg)
else: 
    print("CPF valido", cpf, cpf_sem_dg)        

Cpf invalido 52288166827 22222222242


# 3.3 Logica e operadores bit-a-bit
### operadores logicos and, or e not 
| A | B | A and B | A or B | not A |
| - | - | :-: | :-: | :-: |
| True | True | True | True | False |
| True | False | False | True | False|
| False | True | False | True | True |
| False | False | False | False | True |  

In [39]:
idade = 19 
if idade>=18 and idade<=60:
    print("Pode dirigir sem restrições, desde com carta.")
idade = 10 
while idade<18 or idade>60:
     idade = int(input("Digite uma idade entre 18 e 60:" ))
     if idade<18 or idade>60: print("idade invalida!")
     if not(idade>=18 and idade<=60): print("idade invalida")
# Só vai sair do laço com uma idade entre 18 e 60         

Pode dirigir sem restrições, desde com carta.


In [44]:
idade = int(input("Digite uma idade maior que zero"))
if not idade:
    print("idade invalida!")

idade invalida!


### Operadores bit-abit
#### Operador de conjunção ( & ) ou and bit-a-bit

2 & 3 = 2
2 10 &
3 11
----
2 10 

In [ ]:
a = 5 
b = 7
print(a,f"teste{a:04b}")
print(b,f"{b:04b}","&")
print("-"*7)
print(a&b,f"{(a&b):04b}")
print(bin(a))


5 0101
7 0111 &
-------
5 0101
0b101


#### Operador de disjunção ( | ) ou or bit-a-bit

In [49]:
a = 5 
b = 7
print(a,f"teste{a:04b}")
print(b,f"{b:04b}","&")
print("-"*7)
print(a|b,f"{(a|b):04b}")

5 teste0101
7 0111 &
-------
7 0111


#### Operador xor ( ^ ) ou xor exclusivo bit-a-bit
5 0101
7 0111 ^

------
2 0010

#### Completo ( ~) 
~5 = -6

In [54]:
print(~5)
print(~-6)

-6
5


#### deslocamento de bit para esquerda ( << )
1 << 2 resulta em 100


In [55]:
print(1 << 2, f"{(1<<2):04b}")


4 0100


#### Deslocamento de bit para direita ( >> ) 
5 >> 2 resulta em 1 

In [57]:
print(5 >> 2, f"{(5 >> 2):04b}")

1 0001


### Atualizando a tabela de precedencia de operadores
| Sequencia | Operador | Descrição |
|:-:|:-:| - |
| 1 | ** | Exponenciação |
| 2 | -,+,~ | Operadores unarios e complementos|
| 3 | *,/,//,% | Multiplicação, Divisões e resto |
| 5 | <<, >> | Deslocamento de bits |
| 6 | & | Conjunção |
| 7 | \| | Disjunção |
| 8 | <,<=,>,>=,!=,== | Operadores Relacionais |
| 9 | not | operador de negação |
| 10 | and | Operador and |
| 11 | or | Operador or |
| 12 | +=,-+,*=... | Operador atalho | 